### PR TITLE
Fix double sharding in ShuffleBuffer

### LINKED ISSUE #7986: ShuffleBuffer not returning all patches
**Describe the bug**
While following the tutorial https://github.com/Project-MONAI/tutorials/blob/main/modules/2d_slices_from_3d_training.ipynb and implementing parts of it to my project especially when it comes to transforming Dataset containing 3D to 2D patches I encountered an issue with ShuffleBuffer.

`def create_dataset_2D_ds(ds, keys: list, trans2d: list) -> monai.data.ShuffleBuffer:
    # ds = CacheDataset(data=data_dicts, transform=transforms)
    patch_func = monai.data.PatchIterd(
        keys=keys, patch_size=(None, None, 1), start_pos=(0, 0, 0)
    )

    patch_transform = Compose([
            SqueezeDimd(keys=keys, dim=-1)
            ] + trans2d)

    patch_ds = monai.data.GridPatchDataset( # -> GridPatchDataset contains 531 entries
        data=ds, patch_iter=patch_func, transform=patch_transform, with_coordinates=False
    )
    sb = monai.data.ShuffleBuffer(patch_ds, buffer_size=200, seed=42, epochs=1) # -> ShuffleBuffer returning only 133 of them
    return patch_ds`

The problem is described above in the comments, I know that the GridPatchDataset contains 531 entries, but using Shuffle Buffer in DataLoader will result in 133.

**Expected behavior**
Being able to Iterate with ShuffleBuffer through entire GridPatchDataset

### PR DESCRIPTION
**Pull request by yinkev**
Closes #7986.

### Description

`ShuffleBuffer` no longer applies worker partitioning twice when its source is already a PyTorch/MONAI `IterableDataset`.

A MONAI `IterableDataset` shards its source in its own iterator. `ShuffleBuffer` then wrapped that already-sharded generator in another `IterableDataset`, which sharded it again. With two workers this reduced 40 unique source items to 20 outputs. Ordinary iterable sources retain the existing shuffle-then-partition path, and transforms remain exactly-once.

### Types of changes
- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [x] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### Validation

- deterministic regression fails on `upstream/dev` due to double sharding;
- `python -m pytest tests/data/test_shuffle_buffer.py tests/data/test_iterable_dataset.py -q`: 5 passed;
- real two-worker reproduction: 40 inputs, 40 unique outputs;
- `ruff check` on both touched files;
- `git diff --check`.

### CODE REVIEW

**Comment by yinkev** [comment]
@coderabbitai review

**Comment by yinkev** [comment]
Replaced by #9020 to conform to this repository's documented contributor workflow and branch-naming convention. The replacement uses the same signed commit and identical code diff: https://github.com/Project-MONAI/MONAI/pull/9020
